---
title: "netOmics"
author: "pepe"
date: "4/5/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
Se instalan las librerias necesarias para correr el analisis de netOMics 
```{r,eval=FALSE}
# instalar el paquete por github
library(devtools)
#install_github("abodein/netOmics")
```
```{r, eval=TRUE, message=FALSE}

library(netOmics)
```
```{r, eval=TRUE, message=FALSE}
#if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

#BiocManager::install("timeOmics")

#install_github("zdk123/SpiecEasi")

library(timeOmics)
library(tidyverse)
library(igraph)
library(SpiecEasi)



```

Ahora se cargan los datos del ejemplo de diabetes estacionaria
```{r load_data}
# load data
data("hmp_T2D")
```

Primero se obtienen los dos clusters y las interacciones del RNA por medio del 
algoritmo ARACNe
```{r timeOmics_2}

cluster.info <- hmp_T2D$getCluster.res
```
```{r graph.rna, warning=FALSE}
cluster.info.RNA <- timeOmics::getCluster(cluster.info, user.block = "RNA")
graph.rna <- get_grn(X = hmp_T2D$data$RNA, cluster = cluster.info.RNA)
# informacion de la grafica
get_graph_stats(graph.rna)
```
Ahora se infiere la red y los clusters de las proteinas por medio de la base de
de datos de biogrid

```{r PROT_graph, warning=FALSE}

get_list_mol_cluster <- function(cluster.info, user.block){
  require(timeOmics)
    tmp <- timeOmics::getCluster(cluster.info, user.block) 
    res <- tmp %>% split(.$cluster) %>% 
        lapply(function(x) x$molecule)
    res[["All"]] <- tmp$molecule
    return(res)
}
cluster.info.prot <- get_list_mol_cluster(cluster.info, user.block = 'PROT')
graph.prot <-  get_interaction_from_database(X = cluster.info.prot, 
                                             db = hmp_T2D$interaction.biogrid, 
                                             type = "PROT", user.ego = TRUE)
get_graph_stats(graph.prot)
```
Ahora se hace el procedimieto para los otu de intestino, en este caso se utiliza
el algoritmo sparCC del paquete SpiecEasi
```{r GUT_graph, eval = FALSE}

get_sparcc_graph <- function(X, threshold = 0.3){
    res.sparcc <- sparcc(data = X)
    sparcc.graph <- abs(res.sparcc$Cor) >= threshold
    colnames(sparcc.graph) <-  colnames(X)
    rownames(sparcc.graph) <-  colnames(X)
    res.graph <- graph_from_adjacency_matrix(sparcc.graph, 
                                             mode = "undirected") %>% simplify
    return(res.graph)
}
gut_list <- get_list_mol_cluster(cluster.info, user.block = 'GUT')
graph.gut <- list()
graph.gut[["All"]] <- get_sparcc_graph(hmp_T2D$raw$GUT, threshold = 0.3)
graph.gut[["1"]] <- get_sparcc_graph(hmp_T2D$raw$GUT %>% 
                                       dplyr::select(gut_list[["1"]]), 
                                     threshold = 0.3)
graph.gut[["-1"]] <- get_sparcc_graph(hmp_T2D$raw$GUT %>% 
                                        dplyr::select(gut_list[["-1"]]), 
                                      threshold = 0.3)
class(graph.gut) <- "list.igraph"
```
```{r GUT}
graph.gut <- hmp_T2D$graph.gut
get_graph_stats(graph.gut)
```

Posteirormente se infiere la red de el metaboloma, las variables
clinicas y se construye de biogrid la red de las citocinas 

```{r CYTO_graph, warning=FALSE}
# construccion de las citocinas
cyto_list = get_list_mol_cluster(cluster.info = cluster.info, 
                                 user.block = "CYTO")
graph.cyto <-  get_interaction_from_database(X = cyto_list,
                                             db = hmp_T2D$interaction.biogrid, 
                                             type = "CYTO", user.ego = TRUE)
get_graph_stats(graph.cyto)
# Inferencia del metaboloma
cluster.info.metab <-  timeOmics::getCluster(X = cluster.info, 
                                             user.block = "METAB")
graph.metab <-  get_grn(X = hmp_T2D$data$METAB, 
                        cluster = cluster.info.metab)
get_graph_stats(graph.metab)
# Inferencia clinica
cluster.info.clinical <- timeOmics::getCluster(X = cluster.info, 
                                               user.block = 'CLINICAL')
graph.clinical <- get_grn(X = hmp_T2D$data$CLINICAL,
                          cluster = cluster.info.clinical)
get_graph_stats(graph.clinical)
```
Ahora se combinan las capas creadas con la funcion combine_layers, primero se van
a combinar las capas del proteoma, las citocinas y los FT
```{r, merged_0}
full.graph <- combine_layers(graph1 = graph.rna, graph2 = graph.prot)
full.graph <- combine_layers(graph1 = full.graph, graph2 = graph.cyto)
full.graph <- combine_layers(graph1 = full.graph,
                             graph2 = hmp_T2D$interaction.TF)
 get_graph_stats(full.graph)
```
Ahora se van a unir las capas que no tienen interacciones descritas, sino que 
fueron inferidas, para estas capas se utilizan correlaciones entre los datos 
inferidos y los datos descritos


```{r merged_1_gut, warning=FALSE}
all_data <- reduce(hmp_T2D$data, cbind)
# Microbioma
gut_list <- get_list_mol_cluster(cluster.info, user.block = "GUT")
omic_data <- lapply(gut_list, function(x)dplyr::select(hmp_T2D$data$GUT, x))
# RNA, PROT, CYTO
other_data_list <- get_list_mol_cluster(cluster.info,
                                        user.block = c("RNA", "PROT", "CYTO"))
other_data <- lapply(other_data_list, function(x)dplyr::select(all_data, x))
# Interaccion entre microbioma y RNA, PROT y CYTO
interaction_df_gut <- get_interaction_from_correlation(X = omic_data,
                                                       Y = other_data,
                                                       threshold = 0.99)
# Se une a la grafica completa
full.graph <- combine_layers(graph1 = full.graph,
                             graph2 = graph.gut,
                             interaction.df = interaction_df_gut$All)
```

Se unuen los datos clinicos
```{r, merged_2_clinical, warning=FALSE}
# Datos clinicos
clinical_list <- get_list_mol_cluster(cluster.info, user.block = "CLINICAL")
omic_data <- lapply(clinical_list, 
                    function(x)dplyr::select(hmp_T2D$data$CLINICAL, x))
# otros datos, con la grafica obtenida antes
other_data_list <- get_list_mol_cluster(cluster.info,
                                        user.block = c("RNA", "PROT", 
                                                       "CYTO", "GUT"))
other_data <- lapply(other_data_list, function(x)dplyr::select(all_data, x))
# Interaccion entre los otros datos y los datos clinicos
interaction_df_clinical <- get_interaction_from_correlation(X = omic_data
                                                            , Y = other_data,
                                                            threshold = 0.99)
# Se une con toda la grafica
full.graph <- combine_layers(graph1 = full.graph,
                             graph2 = graph.clinical, 
                             interaction.df = interaction_df_clinical$All)
```

Se Unen los datos del metaboloma
```{r, merged_3_metab, warning=FALSE}
# Metaboloma
metab_list <- get_list_mol_cluster(cluster.info, user.block = "METAB")
omic_data <- lapply(metab_list, function(x)dplyr::select(hmp_T2D$data$METAB, x))
# Los demas datos
other_data_list <- get_list_mol_cluster(cluster.info,
                                        user.block = c("RNA", "PROT", "CYTO", 
                                                       "GUT", "CLINICAL"))
other_data <- lapply(other_data_list, function(x)dplyr::select(all_data, x))
# Interacciones de todos los datos omicos
interaction_df_metab <- get_interaction_from_correlation(X = omic_data,
                                                         Y = other_data, 
                                                         threshold = 0.99)
# Se une todo
full.graph <- combine_layers(graph1 = full.graph, 
                             graph2 = graph.metab, 
                             interaction.df = interaction_df_metab$All)
```


Ahora que se unieron todas las capas se va a hacer el Over Representation Analysis
ORA en donde se introducen los datos de gene ontology GO, pero solo con las capas 
que cuenta con esta informacion, se utiliza la funcion  get_interaction_from_ORA
```{r}
# ORA por cluster
mol_ora <- get_list_mol_cluster(cluster.info, 
                                user.block = c("RNA", "PROT", "CYTO"))
# Se obtienen las interacciones ORA por cluster
graph.go <- get_interaction_from_ORA(query = mol_ora,
                                     sources = "GO",
                                     organism = "hsapiens",
                                     signif.value = TRUE)
# se unen
full.graph <- combine_layers(graph1 = full.graph, graph2 = graph.go)
```

Ademas del analisis ORA, tambien se añadio informacion adicional de medline ranker 
```{r}
# medlineRanker -> datos
medlineranker.res.df <- hmp_T2D$medlineranker.res.df %>% 
  dplyr::select(Disease, symbol) %>% 
  set_names(c("from", "to"))
  
mol_list <-  get_list_mol_cluster(cluster.info = cluster.info,
                                  user.block = c("RNA", "PROT", "CYTO"))
graph.medlineranker <-  get_interaction_from_database(X = mol_list,
                                                      db = medlineranker.res.df, 
                                                      type = "Disease",
                                                      user.ego = TRUE)
get_graph_stats(graph.medlineranker)
# Union
full.graph <- combine_layers(graph1 = full.graph, graph2 = graph.medlineranker)
```
Se completo la preparacion de la red multiomica, tomando en cuenta estas 
caracteristicas. 

* mode = "core" if the vertex was originally present in the data; "extended"
otherwise
* sparse = TRUE if the vertex was present in kinetic cluster signature; FALSE
otherwise
* type = type of omics ("RNA","PROT","CLINICAL","CYTO","GUT","METAB","GO",
"Disease")
* cluster = '1', '-1' or 'NA' (for vertices not originally present in the
original data)
```{r}
# graph cleaning
graph_cleaning <- function(X, cluster.info){
    # no reusability
    X <- igraph::simplify(X)
    va <- vertex_attr(X)
    viewed_mol <- c()
    for(omic in unique(cluster.info$block)){
        mol <- intersect(cluster.info %>% dplyr::filter(.$block == omic) %>%
                           pull(molecule), V(X)$name)
        viewed_mol <- c(viewed_mol, mol)
        X <- set_vertex_attr(graph = X, 
                             name = "type", 
                             index = mol, 
                             value = omic)
        X <- set_vertex_attr(graph = X, 
                             name = "mode",
                             index = mol,
                             value = "core")
    }
    # add medline ranker and go
    mol <- intersect(map(graph.go, ~ as_data_frame(.x)$to) %>%
                       unlist %>% unique(), V(X)$name) # only GO terms
    viewed_mol <- c(viewed_mol, mol)
    X <- set_vertex_attr(graph = X, name = "type", index = mol, value = "GO")
    X <- set_vertex_attr(graph = X, name = "mode", 
                         index = mol, value = "extended")
    
    mol <- intersect(as.character(medlineranker.res.df$from), V(X)$name)
    viewed_mol <- c(viewed_mol, mol)
    X <- set_vertex_attr(graph = X, name = "type",
                         index = mol, value = "Disease")
    X <- set_vertex_attr(graph = X, name = "mode",
                         index = mol, value = "extended")
    
    other_mol <- setdiff(V(X), viewed_mol)
    if(!is_empty(other_mol)){
        X <- set_vertex_attr(graph = X, name = "mode",
                             index = other_mol, value = "extended")
    }
    X <- set_vertex_attr(graph = X, name = "mode", 
                         index = intersect(cluster.info$molecule, V(X)$name), 
                         value = "core")
    
    # signature
    mol <-  intersect(V(X)$name, hmp_T2D$getCluster.sparse.res$molecule)
    X <- set_vertex_attr(graph = X, name = "sparse", index = mol, value = TRUE)
    mol <-  setdiff(V(X)$name, hmp_T2D$getCluster.sparse.res$molecule)
    X <- set_vertex_attr(graph = X, name = "sparse", index = mol, value = FALSE)
    
    return(X)
}
```


```{r}
FULL <- lapply(full.graph, function(x) graph_cleaning(x, cluster.info))
get_graph_stats(FULL)
```
Se pueden ver estadisticas basicas de la red como degree 
distribution, modularity, y short path

```{r, eval = FALSE}
# degree analisis
d <- degree(FULL$All)
hist(d)
d[max(d)]
# modularity 
res.mod <- edge.betweenness.community(FULL$All)
res.mod <- walktrap.community(FULL$All)
# ...
# modularity
sp <- shortest.paths(FULL$All)
```
Ahora sigue el analisis de Random walk, este metodo da redes de 25 nodos a partir 
de una semilla que se selecciona, primero se comienza con las semillas que 
Pertenecen a los elementos GO, para ver las interacciones con ese proceso que no se 
habian descrito en el estudio inicial
```{r}
seeds <- head(V(FULL$All)$name[V(FULL$All)$type == "GO"])
rwr_res <- random_walk_restart(FULL, seeds)
```
Tambien se puede analizar que nodos interactuan con varias capas 
omicas con poca distancia, la funcion "rwr_find_seeds_between_attributes" puede 
encontrar que semillas van a tener mucho atributos con pocas conexiones.
```{r}
rwr_type_k15 <- rwr_find_seeds_between_attributes(X = rwr_res, 
                                                  attribute = "type", k = 15)

summary_plot_rwr_attributes(rwr_type_k15)
summary_plot_rwr_attributes(rwr_type_k15$All)
```
O tambien se pueden buscar nodos que interactuen con diferentes clusters 
```{r}
rwr_type_k15 <- rwr_find_seeds_between_attributes(X = rwr_res$All, 
                                                  attribute = "cluster", k = 15)
summary_plot_rwr_attributes(rwr_type_k15)
```
Tambien se pueden crear subredes de nodos especificos 
```{r}
sub_res <- rwr_type_k15$`GO:0005737`
sub <- plot_rwr_subnetwork(sub_res, legend = TRUE, plot = TRUE)
```

Por ultimo se puede predecir la funcion de un nodo no anotado al hacer el analisis 
random walk con ese nodo como semilla y ver por ejemplo cuales son los 5 elementos
GO mas cercanos a esa semilla para determinar su función. Tambien se puede hacer 
con capas
```{r}
rwr_res <- random_walk_restart(FULL$All, seed = "ZNF263")
# closest GO term
rwr_find_closest_type(rwr_res, seed = "ZNF263", attribute = "type", 
                      value = "GO", top = 5)
# closest Disease
rwr_find_closest_type(rwr_res, seed = "ZNF263", attribute = "type", 
                      value = "Disease", top = 5)
# closest nodes with an attribute "cluster" and the value "-1"
rwr_find_closest_type(rwr_res, seed = "ZNF263", attribute = "cluster",
                      value = "-1", top = 5)
```








